### PR TITLE
Fixing ARO API to enable Terraform Provider for ARO

### PR DESCRIPTION
**Pull request by gouthamMN**
### Which issue this PR addresses:

Fixes https://issues.redhat.com/browse/ARO-2124

### What this PR does / why we need it:

OpenShiftCluster GET API returns enriched worker profile data under a new field "workerProfilesStatus". This change is introduced only to new API. More details can be found on [Design Doc](https://microsoft-my.sharepoint.com/:w:/r/personal/gniranjan_microsoft_com/_layouts/15/doc2.aspx?sourcedoc=%7B920CBF46-6A85-484C-93C9-F7DC3EAF7220%7D&file=OpenShift%20cluster%20-%20Create%20API.docx&action=default&mobileredirect=true&DefaultItemOpen=1&ct=1687552228824&wdOrigin=OFFICECOM-WEB.START.EDGEWORTH&cid=0a2f122b-8dfa-4561-8c90-dc131ad9efc2&wdPreviousSessionSrc=HarmonyWeb&wdPreviousSession=c6aedd5b-f157-4414-aec1-aa5fd19c133e&share=IQFGvwyShWpMSJPJ99w-r3IgAXlNyUSBQnB0VSUWgdbjT5I)

Any API call made to endpoints **/openShiftClusters** or **/openShiftCluster** with GET method using newer API versions (>= v2023-07-01-preview) will return the enriched worker profile data under a new field "**workerProfilesStatus**" in the response and "**workerProfiles**" filed will contain the last input that was passed in PUT/PATCH request body for cluster creation or update.

The response and behavior of any older APIs versions (< v2023-07-01-preview) remain unchanged i.e, only "workerProfiles" filed exists in the response.

### Test plan for issue:

Existing unit tests and E2E test copied over from previous API.

### Is there any documentation that needs to be updated for this PR?

N/A

### CODE REVIEW

**Comment by weinong** [comment]
i think it'd be okay to summarize the outcome from design doc in the PR body. In particular, that table in page 6.

File: pkg/api/admin/openshiftcluster.go [hunk 1]
```diff
@@ -46,6 +46,7 @@ type OpenShiftClusterProperties struct {
 	NetworkProfile                  NetworkProfile          `json:"networkProfile,omitempty"`
 	MasterProfile                   MasterProfile           `json:"masterProfile,omitempty"`
 	WorkerProfiles                  []WorkerProfile         `json:"workerProfiles,omitempty"`
+	WorkerProfilesStatus            []WorkerProfile         `json:"workerProfilesStatus,omitempty"`
```
Comment (weinong) [inline pkg/api/admin/openshiftcluster.go:49]: can we add a comment about this property. in particular, how is it different than the one above it.

File: pkg/api/openshiftcluster.go [hunk 2]
```diff
@@ -128,6 +128,8 @@ type OpenShiftClusterProperties struct {
 
 	WorkerProfiles []WorkerProfile `json:"workerProfiles,omitempty"`
 
+	WorkerProfilesStatus []WorkerProfile `json:"workerProfilesStatus,omitempty"`
```
Comment (weinong) [inline pkg/api/openshiftcluster.go:131]: like previous comment, i think it's better to comment the differences

File: pkg/api/v20191231preview/openshiftcluster_example.go [hunk 3]
```diff
@@ -9,6 +9,7 @@ import (
 
 func exampleOpenShiftCluster() *OpenShiftCluster {
 	doc := api.ExampleOpenShiftClusterDocument()
+	doc.OpenShiftCluster.Properties.WorkerProfilesStatus = nil
```
Comment (weinong) [inline pkg/api/v20191231preview/openshiftcluster_example.go:12]: why is this change needed?

File: pkg/frontend/openshiftcluster_putorpatch.go [hunk 4]
```diff
@@ -252,6 +252,9 @@ func (f *frontend) _putOrPatchOpenShiftCluster(ctx context.Context, log *logrus.
 	doc.OpenShiftCluster.Properties.ClusterProfile.PullSecret = ""
 	doc.OpenShiftCluster.Properties.ServicePrincipalProfile.ClientSecret = ""
 
+	// We don't return enriched worker profile data on PUT/PATCH operations
+	doc.OpenShiftCluster.Properties.WorkerProfilesStatus = nil
```
Comment (weinong) [inline pkg/frontend/openshiftcluster_putorpatch.go:256]: it should be okay to return enriched worker profile since it's not an input

File: pkg/api/v20191231preview/openshiftcluster_example.go (hunk 3 above)
Comment (weinong) [reply pkg/api/v20191231preview/openshiftcluster_example.go:12]: oh i see. it's used in the swagger example generator.

Comment (gouthamMN) [reply pkg/api/v20191231preview/openshiftcluster_example.go:12]: These functions are used by swagger to generate swagger examples (/swagger/redhatopenshift/resource-manager/Microsoft.RedhatOpenShift) of each api versions that we maintain, setting this value to nil will avoid "workerProfilesStatus" filed from swagger examples files. Function **api.ExampleOpenShiftClusterDocument()** returns OC document object defined under pkg/api/openShiftCluster.go, the same is referenced by all api versions, since we want the **workerProfilesStatus** field to be added under new API (v20230701preview) only, we will have to set this to "nil" here for the APIs for which we don't want this field.

File: pkg/frontend/openshiftcluster_putorpatch.go (hunk 4 above)
Comment (gouthamMN) [reply pkg/frontend/openshiftcluster_putorpatch.go:256]: We decided to return the **workerProfilesStatus** field only under GET/LIST operation during design meeting.

File: swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2023-07-01-preview/redhatopenshift.json [hunk 5]
```diff
@@ -2093,6 +2093,14 @@
           },
           "x-ms-identifiers": []
         },
+        "workerProfilesStatus": {
+          "description": "The cluster worker profiles status.",
```
Comment (weinong) [inline swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2023-07-01-preview/redhatopenshift.json:2097]: we should mark this property readonly

**Review by weinong — CHANGES_REQUESTED** [review]
I finished my review.

File: pkg/frontend/openshiftcluster_putorpatch.go (hunk 4 above)
Comment (weinong) [reply pkg/frontend/openshiftcluster_putorpatch.go:256]: okay

**Comment by gouthamMN** [comment]
> i think it'd be okay to summarize the outcome from design doc in the PR body. In particular, that table in page 6.

updated.

**Review by weinong — DISMISSED** [review]
LGTM

File: pkg/api/admin/openshiftcluster_convert.go [hunk 6]
```diff
@@ -212,6 +227,18 @@ func (c openShiftClusterConverter) ToInternal(_oc interface{}, out *api.OpenShif
...
+		for i := range oc.Properties.WorkerProfilesStatus {
+			out.Properties.WorkerProfilesStatus[i].Name = oc.Properties.WorkerProfilesStatus[i].Name
+			out.Properties.WorkerProfilesStatus[i].VMSize = api.VMSize(oc.Properties.WorkerProfilesStatus[i].VMSize)
+			out.Properties.WorkerProfilesStatus[i].DiskSizeGB = oc.Properties.WorkerProfilesStatus[i].DiskSizeGB
+			out.Properties.WorkerProfilesStatus[i].SubnetID = oc.Properties.WorkerProfilesStatus[i].SubnetID
+			out.Properties.WorkerProfilesStatus[i].Count = oc.Properties.WorkerProfilesStatus[i].Count
+			out.Properties.WorkerProfilesStatus[i].EncryptionAtHost = api.EncryptionAtHost(oc.Properties.WorkerProfilesStatus[i].EncryptionAtHost)
+			out.Properties.WorkerProfilesStatus[i].DiskEncryptionSetID = oc.Properties.WorkerProfilesStatus[i].DiskEncryptionSetID
+		}
+	}
```
Comment (facchettos) [inline pkg/api/admin/openshiftcluster_convert.go:241]: it would make more sense to use append here, also creating the object and then putting it in the slice would be much nicer than setting its fields one by one

File: pkg/api/v20230701preview/openshiftcluster_convert.go [hunk 7]
```diff
@@ -161,6 +178,18 @@ func (c openShiftClusterConverter) ToInternal(_oc interface{}, out *api.OpenShif
...
+		for i := range oc.Properties.WorkerProfilesStatus {
+			out.Properties.WorkerProfilesStatus[i].Name = oc.Properties.WorkerProfilesStatus[i].Name
+			out.Properties.WorkerProfilesStatus[i].VMSize = api.VMSize(oc.Properties.WorkerProfilesStatus[i].VMSize)
+			out.Properties.WorkerProfilesStatus[i].DiskSizeGB = oc.Properties.WorkerProfilesStatus[i].DiskSizeGB
+			out.Properties.WorkerProfilesStatus[i].SubnetID = oc.Properties.WorkerProfilesStatus[i].SubnetID
+			out.Properties.WorkerProfilesStatus[i].Count = oc.Properties.WorkerProfilesStatus[i].Count
+			out.Properties.WorkerProfilesStatus[i].EncryptionAtHost = api.EncryptionAtHost(oc.Properties.WorkerProfilesStatus[i].EncryptionAtHost)
+			out.Properties.WorkerProfilesStatus[i].DiskEncryptionSetID = oc.Properties.WorkerProfilesStatus[i].DiskEncryptionSetID
+		}
+	}
```
Comment (facchettos) [inline pkg/api/v20230701preview/openshiftcluster_convert.go:192]: same comment as above

**Review by facchettos — CHANGES_REQUESTED** [review]

**Review by s-amann — APPROVED** [review]

**Comment by SudoBrendan** [comment]
Hey, I'm adding a `hold` on this since it looks like we might be changing up what API version this lands in. Please reach out if you have any questions.
